src/Transport/ResponseCorrelator.cs

```diff
@@ -85,15 +87,40 @@ public override FutureResponse AsyncRequest(Command command)
 			return future;
         }
 
-        public override Response Request(Command command, TimeSpan timeout)
+        public override Task<Response> RequestAsync(Command command, TimeSpan timeout)
         {
+	        TaskCompletionSource<Response> taskCompletionSource = new TaskCompletionSource<Response>(TaskCreationOptions.RunContinuationsAsynchronously);
+			if (timeout.TotalMilliseconds > 0)
+	        {
+		        CancellationTokenSource ct = new CancellationTokenSource(timeout);
+		        ct.Token.Register(() =>
+		        {
+			        taskCompletionSource.TrySetException(new RequestTimedOutException(timeout));
+		        }, false);
+	        }
+    
             FutureResponse future = AsyncRequest(command);
```

Comment:
This `FutureResponse` is some Java variation around `TaskCompletionSource`, I'd drop it altogether. Is doesn't add any value, but complicates logic dramatically.